hudi-client/src/main/java/org/apache/hudi/table/action/clustering/HoodieCopyOnWriteTableCluster.java

```diff
@@ -0,0 +1,243 @@
+        p -> new Path(FSUtils.getPartitionPath(metaClient.getBasePath(), operation.getPartitionPath()), p).toString())
+        .collect(toList());
+    List<HoodieFileReader> list = baseFiles.stream().map(s -> {
+      try {
+        return HoodieFileReaderFactory.getFileReader(hoodieCopyOnWriteTable.getHadoopConf(), new Path(s));
+      } catch (IOException e) {
+        throw new HoodieException(e);
+      }
+    }).collect(toList());
+
+    BaseFileIterator baseFileIterator = new BaseFileIterator(list, readerSchema);
+    Iterator<List<WriteStatus>> result = hoodieCopyOnWriteTable.handleInsert(instantTime, operation.getPartitionPath(), FSUtils.createNewFileIdPfx(),
+            baseFileIterator);
+
+    Iterable<List<WriteStatus>> resultIterable = () -> result;
+    return StreamSupport.stream(resultIterable.spliterator(), false).flatMap(Collection::stream).peek(s -> {
+      s.getStat().setPartitionPath(operation.getPartitionPath());
+      RuntimeStats runtimeStats = new RuntimeStats();
+      s.getStat().setRuntimeStats(runtimeStats);
+    }).collect(toList());
+  }
+
+  @Override
+  public HoodieClusteringPlan generateClusteringPlan(JavaSparkContext jsc, HoodieTable hoodieTable,
+                                                     HoodieWriteConfig config, String clusteringCommitTime, Set<HoodieFileGroupId> fgIdsInPendingClusterings)
+      throws IOException {
+
+
+    totalFileSlices = new LongAccumulator();
+    jsc.sc().register(totalFileSlices);
+
+    HoodieTableMetaClient metaClient = hoodieTable.getMetaClient();
+    LOG.info("Clustering " + metaClient.getBasePath() + " with commit " + clusteringCommitTime);
+    List<String> partitionPaths = FSUtils.getAllPartitionPaths(metaClient.getFs(), metaClient.getBasePath(),
+        config.shouldAssumeDatePartitioning());
+
+    // filter the partition paths if needed to reduce list status
+    partitionPaths = config.getClusteringStrategy().filterPartitionPaths(config, partitionPaths);
+
+    if (partitionPaths.isEmpty()) {
+      // In case no partitions could be picked, return no compaction plan
+      return null;
+    }
+
+    SliceView fileSystemView = hoodieTable.getSliceView();
+    LOG.info("Clustering looking for files to cluster in " + partitionPaths + " partitions");
+    Set<HoodieFileGroupId> fgIdsPendingCompactions = ((SyncableFileSystemView) fileSystemView).getPendingCompactionOperations()
+            .map(instantTimeOpPair -> instantTimeOpPair.getValue().getFileGroupId())
+            .collect(Collectors.toSet());
+    List<HoodieClusteringOperation> operations = jsc.parallelize(partitionPaths, partitionPaths.size()).map((Function<String, ClusteringOperation>) partitionPath -> {
+      Stream<FileSlice> fileSliceStream = fileSystemView.getLatestFileSlices(partitionPath);
+      List<HoodieBaseFile> baseFiles = fileSliceStream.filter(slice -> (!fgIdsInPendingClusterings.contains(slice.getFileGroupId()) && !fgIdsPendingCompactions.contains(slice.getFileGroupId())))
```

Comment:
For COW table, pending compactions would not be present.

Reply:
Some of the code seems common with Compaction handling. Any chance of moving to common helper functions ?

Reply:
thanks, will reuse these code